### ISSUE TITLE
[test_syslog_rate_limit] 'systemctl status rsyslog' returned non-zero exit status 3

### ISSUE THREAD

**Issue by lipxu** [sonic-net/sonic-mgmt#10541]
**Description**

**Steps to reproduce the issue:**
1. Run case "syslog/test_syslog_rate_limit.py::test_syslog_rate_limit", failed due to Command 'systemctl status rsyslog' returned non-zero exit status 3."
3.
4.

**Describe the results you received:**
2023-05-10T23:43:38.3308139Z collected 1 item
2023-05-10T23:43:38.3309577Z 
2023-05-10T23:45:27.5847757Z syslog/test_syslog_rate_limit.py::test_syslog_rate_limit 
2023-05-10T23:45:27.5853821Z -------------------------------- live log call ---------------------------------
2023-05-10T23:45:27.5857753Z 23:45:27 __init__.pytest_runtest_call             L0040 ERROR  | Traceback (most recent call last):
2023-05-10T23:45:27.5862099Z   File "/var/AzDevOps/env-python3/lib/python3.8/site-packages/_pytest/python.py", line 1761, in runtest
2023-05-10T23:45:27.5865206Z     self.ihook.pytest_pyfunc_call(pyfuncitem=self)
2023-05-10T23:45:27.5868939Z   File "/var/AzDevOps/env-python3/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
2023-05-10T23:45:27.5872173Z     return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
2023-05-10T23:45:27.5876119Z   File "/var/AzDevOps/env-python3/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
2023-05-10T23:45:27.5879329Z     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
2023-05-10T23:45:27.5883166Z   File "/var/AzDevOps/env-python3/lib/python3.8/site-packages/pluggy/_callers.py", line 60, in _multicall
2023-05-10T23:45:27.5886218Z     return outcome.get_result()
2023-05-10T23:45:27.5889540Z   File "/var/AzDevOps/env-python3/lib/python3.8/site-packages/pluggy/_result.py", line 60, in get_result
2023-05-10T23:45:27.5893182Z     raise ex[1].with_traceback(ex[2])
2023-05-10T23:45:27.5896566Z   File "/var/AzDevOps/env-python3/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
2023-05-10T23:45:27.5899532Z     res = hook_impl.function(*args)
2023-05-10T23:45:27.5903032Z   File "/var/AzDevOps/env-python3/lib/python3.8/site-packages/_pytest/python.py", line 192, in pytest_pyfunc_call
2023-05-10T23:45:27.5906204Z     result = testfunction(**testargs)
2023-05-10T23:45:27.5909192Z   File "/azp/_work/4/s/tests/syslog/test_syslog_rate_limit.py", line 78, in test_syslog_rate_limit
2023-05-10T23:45:27.5912218Z     verify_host_rate_limit(rand_selected_dut)
2023-05-10T23:45:27.5915246Z   File "/azp/_work/4/s/tests/syslog/test_syslog_rate_limit.py", line 186, in verify_host_rate_limit
2023-05-10T23:45:27.5918379Z     verify_rate_limit_with_log_generator(rand_selected_dut,
2023-05-10T23:45:27.5921956Z   File "/azp/_work/4/s/tests/syslog/test_syslog_rate_limit.py", line 243, in verify_rate_limit_with_log_generator
2023-05-10T23:45:27.5924871Z     with loganalyzer:
2023-05-10T23:45:27.5927714Z   File "/azp/_work/4/s/tests/common/plugins/loganalyzer/loganalyzer.py", line 120, in __enter__
2023-05-10T23:45:27.5930657Z     self._markers.append(self.init())
2023-05-10T23:45:27.5933662Z   File "/azp/_work/4/s/tests/common/plugins/loganalyzer/loganalyzer.py", line 244, in init
2023-05-10T23:45:27.5936688Z     return self._setup_marker(log_files=log_files)
2023-05-10T23:45:27.5940032Z   File "/azp/_work/4/s/tests/common/plugins/loganalyzer/loganalyzer.py", line 284, in _setup_marker
2023-05-10T23:45:27.5942979Z     self.ansible_host.command(cmd)
2023-05-10T23:45:27.5945817Z   File "/azp/_work/4/s/tests/common/devices/multi_asic.py", line 118, in _run_on_asics
2023-05-10T23:45:27.5949272Z     return getattr(self.sonichost, self.multi_asic_attr)(*module_args, **complex_args)
2023-05-10T23:45:27.5952376Z   File "/azp/_work/4/s/tests/common/devices/base.py", line 127, in _run
2023-05-10T23:45:27.5955701Z     raise RunAnsibleModuleFail("run module {} failed".format(self.module_name), res)
2023-05-10T23:45:27.5959159Z tests.common.errors.RunAnsibleModuleFail: run module command failed, Ansible Results =>
2023-05-10T23:45:27.5980281Z {"changed": true, "cmd": ["python", "/tmp/loganalyzer.py", "--action", "init", "--run_id", "syslog_rate_limit_host_interval_10_burst_100.2023-05-10-23:45:26"], "delta": "0:00:00.163658", "end": "2023-05-10 23:45:27.555520", "failed": true, "msg": "non-zero return code", "rc": 1, "start": "2023-05-10 23:45:27.391862", "stderr": "Traceback (most recent call last):\n  File \"/tmp/loganalyzer.py\", line 852, in <module>\n    main(sys.argv[1:])\n  File \"/tmp/loganalyzer.py\", line 805, in main\n    analyzer.place_marker(log_file_list, analyzer.create_start_marker())\n  File \"/tmp/loganalyzer.py\", line 254, in place_marker\n    self.place_marker_to_syslog(marker)\n  File \"/tmp/loganalyzer.py\", line 203, in place_marker_to_syslog\n    self.flush_rsyslogd()\n  File \"/tmp/loganalyzer.py\", line 167, in flush_rsyslogd\n    out = str(subprocess.check_output(\"systemctl status rsyslog\", shell=True))\n  File \"/usr/lib/python3.9/subprocess.py\", line 424, in check_output\n    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,\n  File \"/usr/lib/python3.9/subprocess.py\", line 528, in run\n    raise CalledProcessError(retcode, process.args,\nsubprocess.CalledProcessError: Command 'systemctl status rsyslog' returned non-zero exit status 3.", "stderr_lines": ["Traceback (most recent call last):", "  File \"/tmp/loganalyzer.py\", line 852, in <module>", "    main(sys.argv[1:])", "  File \"/tmp/loganalyzer.py\", line 805, in main", "    analyzer.place_marker(log_file_list, analyzer.create_start_marker())", "  File \"/tmp/loganalyzer.py\", line 254, in place_marker", "    self.place_marker_to_syslog(marker)", "  File \"/tmp/loganalyzer.py\", line 203, in place_marker_to_syslog", "    self.flush_rsyslogd()", "  File \"/tmp/loganalyzer.py\", line 167, in flush_rsyslogd", "    out = str(subprocess.check_output(\"systemctl status rsyslog\", shell=True))", "  File \"/usr/lib/python3.9/subprocess.py\", line 424, in check_output", "    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,", "  File \"/usr/lib/python3.9/subprocess.py\", line 528, in run", "    raise CalledProcessError(retcode, process.args,", "subprocess.CalledProcessError: Command 'systemctl status rsyslog' returned non-zero exit status 3."], "stdout": "", "stdout_lines": []}
2023-05-10T23:45:27.5986329Z 
2023-05-10T23:46:04.0686839Z FAILED                                                                   [100%]
2023-05-10T23:46:04.0689357Z 
2023-05-10T23:46:04.0692231Z =================================== FAILURES ===================================
2023-05-10T23:46:04.0695719Z ____________________________ test_syslog_rate_limit ____________________________
2023-05-10T23:46:04.0697096Z 
2023-05-10T23:46:04.0699705Z rand_selected_dut = <MultiAsicSonicHost bjw-can-2700-1>
2023-05-10T23:46:04.0700907Z 
2023-05-10T23:46:04.0702859Z     @pytest.mark.disable_loganalyzer
2023-05-10T23:46:04.0705061Z     def test_syslog_rate_limit(rand_selected_dut):
2023-05-10T23:46:04.0707531Z         """Test case for syslog rate limit
2023-05-10T23:46:04.0709451Z     
2023-05-10T23:46:04.0711230Z         Args:
2023-05-10T23:46:04.0713346Z             rand_selected_dut (object): DUT host object
2023-05-10T23:46:04.0715577Z         """
2023-05-10T23:46:04.0717589Z         # Copy tests/syslog/log_generator.py to DUT
2023-05-10T23:46:04.0720570Z         rand_selected_dut.copy(src=LOCAL_LOG_GENERATOR_FILE, dest=REMOTE_LOG_GENERATOR_FILE)
2023-05-10T23:46:04.0723087Z     
2023-05-10T23:46:04.0725120Z         verify_container_rate_limit(rand_selected_dut)
2023-05-10T23:46:04.0727635Z >       verify_host_rate_limit(rand_selected_dut)

**Describe the results you expected:**
case pass

**Additional information you deem important:**

    **Output of `show version`:**
current image: master.269513-a738c3932

    ```
    (paste your output here)
    ```

    **Attach debug file `sudo generate_dump`:**

    ```
    (paste your output here)
    ```

**Comment by bingwang-ms** [comment]
@roy-sror Can you please help take a look?

**Comment by lipxu** [comment]
@roy-sror may you share the latest update or plan for this issue, thanks a lot.

**Comment by roy-sror** [comment]
@echuawu  - can you pls update with your findings.

**Comment by echuawu** [comment]
I run the script at MSFT image SONiC.20230531.06, and everything works fine, script could pass. Judge from the rc code of 'systemctl status rsyslog' in the Issue, it seems that the rsyslog process is not running due to some reason. It should be a basis for DUT to work properly. Please check the reason of rsyslog failed to work properly. If the rsyslog could works fine, we could debug the script failure if the script still could not pass then.

**Comment by lipxu** [comment]
Hi, @echuawu , thanks for your update. Based on our nightly test result, seems the issue failed more easily on device 2700. Please have a try. Thanks.

**Comment by bingwang-ms** [comment]
I analyzed the syslog on one SN2700 device. It looks like a timing issue to me.
The timeline is as below.
**Nov 30 15:29:19.469866** Call CLI to update rsyslog limit
**Nov 30 15:29:22.680790** Stopping System Logging Service...
**But Stopped System Logging Service and Started Logging Service logs were missing.**
**Nov 30 15:29:26.367040** Call LogAnalyzer to analyze log, then the error happened.

So, probably add a check after changing syslog limit to ensure the service is fully restarted can fix the issue.
 
```
Nov 30 15:29:17.492634 str-msn2700-02 INFO python[508045]: ansible-ansible.legacy.command Invoked with _raw_params=config syslog rate-limit-host -b 100 -i 10 _uses_shell=False warn=False stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None
Nov 30 15:29:19.469866 str-msn2700-02 NOTICE sniffer: Configured syslog host rate-limits: interval=10,        burst=100
Nov 30 15:29:19.479534 str-msn2700-02 INFO hostcfgd: SYSLOG_CONFIG table handler...
Nov 30 15:29:19.520940 str-msn2700-02 DEBUG hostcfgd: RSyslogCfg: Configuration update
Nov 30 15:29:19.521087 str-msn2700-02 INFO hostcfgd: RSyslogCfg: Set config {'GLOBAL': {'rate_limit_burst': '100', 'rate_limit_interval': '10'}}, servers: {'10.64.246.95': {}}
Nov 30 15:29:19.575185 str-msn2700-02 INFO systemd[1]: rsyslog-config.service: Succeeded.
Nov 30 15:29:19.585058 str-msn2700-02 INFO systemd[1]: Stopped Update rsyslog configuration.
Nov 30 15:29:19.585289 str-msn2700-02 INFO systemd[1]: Stopping Update rsyslog configuration...
Nov 30 15:29:19.585581 str-msn2700-02 INFO systemd[1]: Starting Update rsyslog configuration...
Nov 30 15:29:22.676995 str-msn2700-02 INFO systemd[1]: Stopping Host core file uploader daemon...
Nov 30 15:29:22.677178 str-msn2700-02 INFO systemd[1]: core_uploader.service: Succeeded.
Nov 30 15:29:22.677292 str-msn2700-02 INFO systemd[1]: Stopped Host core file uploader daemon.
Nov 30 15:29:22.680790 str-msn2700-02 INFO systemd[1]: Stopping System Logging Service...
Nov 30 15:29:22.695674 str-msn2700-02 INFO rsyslogd:  message repeated 3 times: [[origin software="rsyslogd" swVersion="8.2302.0" x-pid="502265" x-info="https://www.rsyslog.com"] rsyslogd was HUPed]
Nov 30 15:29:26.367040 str-msn2700-02 INFO python[508345]: ansible-ansible.legacy.command Invoked with _raw_params=python /tmp/loganalyzer.py --action init --run_id syslog_rate_limit_host_interval_10_burst_100.2023-11-30-15:29:25 _uses_shell=False warn=False stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None
Nov 30 15:29:26.597206 str-msn2700-02 INFO start-LogAnalyzer-syslog_rate_limit_host_interval_10_burst_100.2023-11-30-15:29:25
Nov 30 15:29:26.597376 str-msn2700-02 INFO
Nov 30 15:29:27.780314 str-msn2700-02 INFO python[508375]: ansible-ansible.legacy.command Invoked with _raw_params=config syslog rate-limit-container acms -b 20000 -i 300 _uses_shell=False warn=False stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None
Nov 30 15:29:28.907083 str-msn2700-02 NOTICE sniffer: Configured syslog acms rate-limits: interval=300,        burst=20000
Nov 30 15:29:29.975323 str-msn2700-02 INFO python[508395]: ansible-ansible.legacy.command Invoked with _raw_params=config syslog rate-limit-container bgp -b 20000 -i 300 _uses_shell=False warn=False stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None
Nov 30 15:29:31.196301 str-msn2700-02 NOTICE sniffer: Configured syslog bgp rate-limits: interval=300,        burst=20000
Nov 30 15:29:32.124175 str-msn2700-02 INFO python[508403]: ansible-ansible.legacy.command Invoked with _raw_params=config syslog rate-limit-container database -b 20000 -i 300 _uses_shell=False warn=False stdin_add_newline=True strip_empty_ends=True argv=None chdir=None executable=None creates=None removes=None stdin=None
Nov 30 15:29:33.389432 str-msn2700-02 NOTICE sniffer: Configured syslog database rate-limits: interval=300,        burst=20000
Nov 30 15:29:34.218982 str-msn2700-02 INFO systemd[1]: rsyslog.service: Succeeded.
Nov 30 15:29:34.219137 str-msn2700-02 INFO systemd[1]: Stopped System Logging Service.
Nov 30 15:29:34.223572 str-msn2700-02 INFO systemd[1]: Starting System Logging Service...
Nov 30 15:29:34.243480 str-msn2700-02 INFO systemd[1]: Started System Logging Service.
Nov 30 15:29:34.248855 str-msn2700-02 INFO systemd[1]: Started Host core file uploader daemon.
Nov 30 15:29:34.269102 str-msn2700-02 INFO systemd[1]: Finished Update rsyslog configuration.
```

**Comment by echuawu** [comment]
Hi @bingwang-ms, I will investigate the solution soon.

**Comment by lipxu** [comment]
Hi, @echuawu ,  may you share the latest update or plan for this issue, thanks a lot.

**Comment by echuawu** [comment]
Add Nana in loop.

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by echuawu** [comment]
Hi @bingwang-ms , Junchao had already added a check as you mentioned in https://github.com/sonic-net/sonic-mgmt/pull/7386

![image](https://github.com/sonic-net/sonic-mgmt/assets/103085864/99f88065-c117-41f4-9822-5841db60a0c0)

**Comment by lipxu** [comment]
https://github.com/sonic-net/sonic-mgmt/pull/10867

**Comment by bingwang-ms** [comment]
@lipxu Please double check if the issue is fixed by PR#10867. Thanks!

**Comment by lipxu** [comment]
> @lipxu Please double check if the issue is fixed by PR#10867. Thanks!

Based on the latest nightly run result, I think the issue has been fixed by PR#10867. let me close the issue, thanks.
